### PR TITLE
Change status package to deal with concrete types instead of interfaces

### PR DESCRIPTION
**Pull request by dfawley**

### CODE REVIEW

File: status/status.go [hunk 1]
```diff
@@ -50,78 +50,56 @@ import (
...
 
-func (se *statusError) Err() error {
-	return se
+// Err returns an immutable error representing s; returns nil if s.Code() is
+// OK.
+func (s *Status) Err() error {
+	if s.Code() == codes.OK {
+		return nil
+	}
+	return (*statusError)(proto.Clone(s.s).(*spb.Status))
```
Comment (menghanl) [inline status/status.go]: Why do we still make a clone here?

File: transport/transport.go [hunk 2]
```diff
@@ -335,7 +335,7 @@ func (s *Stream) Read(p []byte) (n int, err error) {
 
 // finish sets the stream's state and status, and closes the done channel.
 // s.mu must be held by the caller.
-func (s *Stream) finish(st status.Status) {
+func (s *Stream) finish(st *status.Status) {
```
Comment (menghanl) [inline transport/transport.go]: Comment this should never be called with nil?

File: status/status.go (hunk 1 above)
Comment (dfawley) [inline status/status.go]: To make it impossible to mutate the statusError and change the Code into OK.

Without this, but the rest of the code as-is, one could do:

status := status.New(codes.Internal, "blah")
p := status.Proto()
err := status.Err()
p.Code = int32(codes.OK)

Maybe status.Proto() and FromProto() should do the clone instead, making Status immutable, which is how I had it initially.

File: transport/transport.go (hunk 2 above)
Comment (dfawley) [inline transport/transport.go]: Done

**Pull request merged**